src/cache/inc/cacheItem.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2019 TAOS Data, Inc. <cli@taosdata.com>
+ *
+ * This program is free software: you can use, redistribute, and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3
+ * or later ("AGPL"), as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef TDENGINE_CACHE_ITEM_H
+#define TDENGINE_CACHE_ITEM_H
+
+#include <stdint.h>
+#include <string.h>
+#include "cachePriv.h"
+#include "cacheTable.h"
+#include "cacheTypes.h"
+#include "osAtomic.h"
+#include "osDef.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum cacheItemFlag {
+  /* item in use state: */
+  ITEM_IN_USE   = 1,  /* item in use, 0 means item is free */
+  ITEM_FETCHED  = 2,  /* item in use and was fetched at least once in its lifetime */
+  ITEM_ACTIVE   = 4,  /* Appended on fetch, removed on LRU shuffling */
+  ITEM_CHUNKED  = 8,  /* item in use and in chunked mode */
+};
+
+struct cacheItem {
+  /* Protected by slab/slab-LRU mutex */
+  struct cacheItem* next;
+  struct cacheItem* prev;
+
+  struct cacheItem* h_next;       /* hash tabel next item, protected by table mutex */
+
+  cacheTable*       pTable;       /* owner cache table */
+  uint16_t          flags;        /* item flags above */
+
+  uint8_t           slabLruId;    /* which slab lru class we're in */
+
+  uint64_t          expireTime;   /* expire time, 0 means no expire time */
+  uint64_t          lastTime;     /* last access time */
+
+  uint8_t           refCnt;
+
+  uint8_t           nkey;         /* key length */
+  
+  uint32_t          nbytes;       /* size of data */
+
+  char              data[];
```

Comment:
Is key also stored in "data" ?

Reply:
Oh, I saw it. Yes. 
#define item_key(item) (((char*)&((item)->data) + (item)->nbytes + 1))
Is the "1" a left blank? Or something like flags?